### PR TITLE
fatal error LNK1235: corrupt or invalid COFF symbol table - Fix Issue 14201

### PR DESCRIPTION
**Pull request by etcimon**
When debugging [this](https://issues.dlang.org/show_bug.cgi?id=14201), I found about 20k sections for `.debug$S` in the COFF table. My program has 50k sections so these are taking all the space when they shouldn't be. I found that they were being forcefully created in `getsegment` even for release builds, whereas the function should have simply returned the index to the section ID, which is always 1 for the COFF format.

Thus, I removed a flag `IMAGE_SCN_LNK_COMDAT` that forced the creation of redundant `.debug$S` sections.

Additionally, I added an assertion test to avoid others to run into this error, because it should have thrown a `too_many_symbols` error rather than spit out a corrupt COFF symbol table.

I pulled the maximum number of sections from here: http://llvm.org/docs/doxygen/html/Support_2COFF_8h_source.html

There is no simple test case for this due to the nature of the error being caused by large numbers of symbols. However, you can experience it by compiling [Botan](https://github.com/etcimon/botan) on win64 with `dub test --arch=x86_64`

The patched compiler successfully runs all cryptography tests with druntime, phobos and botan compiled anew on win64, rather than failing with the linker error.

### CODE REVIEW

**Comment by WalterBright** [comment]
When doing a PR to fix a bugzilla issue, include the phrase:
```
fix Issue NNNNN ...
```
in the title, please, as the tooling relies on that.

File: src/backend/cgobj.c [hunk 1]
```diff
@@ -373,6 +373,9 @@ void too_many_symbols()
 {
 #if SCPP
     err_fatal(EM_too_many_symbols, 0x7FFF);
+#elif TARGET_WINDOS // COFF
```
Comment (WalterBright) [inline src/backend/cgobj.c]: cgobj.c generates OMF, not COFF. Please remove these added lines.

File: src/backend/mscoffobj.c [hunk 2]
```diff
@@ -1602,6 +1604,7 @@ IDXSEC MsCoffObj::addScnhdr(const char *scnhdr_name, unsigned long flags)
         memcpy(sec.s_name, scnhdr_name, len);
     sec.s_flags = flags;
     ScnhdrBuf->write((void *)&sec, sizeof(sec));
+    if (scnhdr_cnt > 65279) too_many_symbols();
```
Comment (WalterBright) [inline src/backend/mscoffobj.c]: add the error message here

**Comment by MartinNowak** [comment]
Sure there is no way to support that many sections?
I had to fix this for ELF as well #1540, there was something called extended section headers.

File: src/backend/cgobj.c [hunk 3]
```diff
@@ -373,6 +373,9 @@ void too_many_symbols()
...
+    error(NULL, 0, 0, "more than %d sections in object file", 65279);
```
Comment (MartinNowak) [inline src/backend/cgobj.c]: What's that number? Google wouldn't find anything with this number and OMF or COFF.

Comment (etcimon) [inline src/backend/cgobj.c]: That number was on http://llvm.org/docs/doxygen/html/Support_2COFF_8h_source.html

**Comment by justanotheranonymoususer** [comment]
Seems like there's something along the lines:
http://reviews.llvm.org/D5349

**Comment by etcimon** [comment]
> Sure there is no way to support that many sections?
> I had to fix this for ELF as well #1540, there was something called extended section headers.

Thanks @justanotheranonymoususer ; yes BigObj supports 2147483647 sections and it's supported by microsoft's linker according to this: https://sourceware.org/ml/binutils/2014-03/msg00110.html

**Comment by MartinNowak** [comment]
How about a separate follow-up pull request, that implements bigobj for COFF?
I actually used it several years ago, didn't remember if exists.

**Comment by etcimon** [comment]
Sure, although I think 16 bit COFF should Be dropped altogether considering 32 bit COFF (bigobj) was implemented in vs 2015. https://msdn.microsoft.com/en-us/library/ms173499(v=VS.100).aspx

It not only needs to be default, but also exclusive.

**Comment by etcimon** [comment]
I'm going to replace the error and symbol limit with the bigobj implementation if you don't mind

**Comment by etcimon** [comment]
done!

**Comment by WalterBright** [comment]
> if you don't mind

Nope, don't mind at all. I didn't know about bigobj!

**Comment by WalterBright** [comment]
Auto-merge toggled on

File: src/backend/mscoffobj.c [hunk 4]
```diff
@@ -614,17 +614,19 @@ void build_syment_table()
 
         if (psechdr->s_flags & IMAGE_SCN_LNK_COMDAT)
         {
-            aux.x_section.Selection = IMAGE_COMDAT_SELECT_ANY;
+            aux.x_section.Selection = (unsigned char)IMAGE_COMDAT_SELECT_ANY;
             if (pseg->SDassocseg)
-            {   aux.x_section.Selection = IMAGE_COMDAT_SELECT_ASSOCIATIVE;
-                aux.x_section.Number = pseg->SDassocseg;
+            {   aux.x_section.Selection = (unsigned char)IMAGE_COMDAT_SELECT_ASSOCIATIVE;
+                aux.x_section.NumberHighPart = (unsigned short)pseg->SDassocseg >> 16;
```
Comment (rainers) [inline src/backend/mscoffobj.c]: Are you sure about this line? It always writes 0 AFAICS.

Comment (Orvid) [inline src/backend/mscoffobj.c]: Also, I believe the current standard is to remove this style of formatting when it's encountered.

Comment (etcimon) [inline src/backend/mscoffobj.c]: If the type cast binds stronger than the right shift, so this line is not alright. The linker worked on my project though. I guess there should be parenthesis here

Comment (etcimon) [inline src/backend/mscoffobj.c]: I submitted a fix here: https://github.com/D-Programming-Language/dmd/pull/4443

**Comment by rainers** [comment]
The bigobj format is not supported by some tools, e.g. objconv by Agner Fog and objdump from the MinGW distribution and only very recently by the LLVM objdump. Most annoying for me at the moment: even dumpbin from VS2013 fails to display symbol labels in the disassembly for big object files.

Can we make the switch to bigobj format optional somehow?

**Comment by rainers** [comment]
> e.g. objconv by Agner Fog and objdump from the MinGW distribution

DM obj2asm does not support bigobj files, too.

**Comment by etcimon** [comment]
This is currently necessary to compile my projects. Not sure if allowing projects to compile would be of higher priority ;)

Maybe we could make the legacy coff a compiler flag?

**Comment by yebblies** [comment]
Oh, that explains why objconv stopped working.  Can we make it so bigobj is only used when the max symbols for normal coff is exceeded?

**Comment by etcimon** [comment]
> Oh, that explains why objconv stopped working. Can we make it so bigobj is only used when the max symbols for normal coff is exceeded?

From how I remember it, all the symbols are being streamed and a placeholder for the file header is used while the symbols are being written to the file, which would make this "fallback" structurally impossible. I think a legacy flag would be a better option

**Comment by yebblies** [comment]
Impossible is a strong word.  We could always count the symbols before doing object file generation.

**Comment by etcimon** [comment]
> Impossible is a strong word. We could always count the symbols before doing object file generation.

I agree. I have a bunch of urgent HTTP/2 bugs to fix right now though so I can't really tackle this

**Comment by rainers** [comment]
https://github.com/D-Programming-Language/dmd/pull/4648

**Pull request merged**